apps/asap-server/src/handlers/webhooks/webhook-run-migrations.ts

```diff
@@ -0,0 +1,257 @@
+    } catch (error) {
+      logger.error(
+        error,
+        `Error executing the rollback of migration ${migration.getPath()}`,
+      );
+      return;
+    }
+
+    logger.debug('Finished executing rollback, saving progress...');
+
+    await removeExecutedMigration(migrationPath);
+
+    logger.info(`Rolled back and removed migration '${migration.getPath()}'`);
+  }
+};
+
+const getMigrationPathsFromDirectoryFactory = (
+  readDir: typeof fsPromise.readdir,
+) => async () => (await readDir(migrationDir)).sort();
+
+const getLatestMigrationPathFromDbFactory = (
+  client: Squidex<RestMigration>,
+) => async (): Promise<string | null> => {
+  const query: Query = {
+    take: 1,
+    skip: 0,
+    sort: [{ path: 'data.name.iv', order: 'descending' }],
+  };
+
+  const { items: migrations } = await client.fetch(query);
+
+  if (migrations.length === 0) {
+    return null;
+  }
+
+  return migrations[0].data.name.iv;
+};
+
+const importModuleFromPath = (path: string): Promise<Module> => import(path);
+export type ImportModuleFromPath = typeof importModuleFromPath;
+
+type Module = {
+  default?: { new (path: string): Migration | unknown };
+};
+
+const getMigrationsFromPathsFactory = (
+  importModule: typeof importModuleFromPath,
+  logger: Logger,
+) => async (migrationPaths: string[]): Promise<Migration[]> => {
+  const migrations = Promise.all(
+    migrationPaths.map(async (file) => {
+      logger.debug({ file });
+      const { default: ImportedModule } = await importModule(
+        `${migrationDir}/${file}`,
+      );
+
+      if (typeof ImportedModule !== 'function') {
+        throw new Error(`${file} does not export a valid module`);
+      }
+
+      const migration = new ImportedModule(file);
+
+      if (!isMigration(migration)) {
+        throw new Error(`${file} does not contain a valid migration`);
+      }
+
+      return migration;
+    }),
+  );
+
+  return migrations;
+};
+
+const filterUnexecutedMigrationsFactory = (client: Squidex<RestMigration>) =>
+  filterMigrationsFactory(client);
+
+const filterMigrationsFactory = (
+  client: Squidex<RestMigration>,
+  unexecuted = true,
+) => (migrationPaths: string[]) => {
+  const asyncFilter = async <T>(
+    arr: T[],
+    predicate: (elem: T) => Promise<boolean>,
+  ): Promise<T[]> => {
+    const results = await Promise.all(arr.map(predicate));
+
+    return arr.filter((_v, index) => results[index]);
+  };
+
+  return asyncFilter(migrationPaths, async (migration) => {
+    try {
+      await client.fetchOne({
+        filter: { path: 'data/name/iv', op: 'eq', value: migration },
+      });
+
+      return !unexecuted;
+    } catch (error) {
+      if (isBoom(error, 404)) {
+        return unexecuted;
+      }
+
+      throw error;
+    }
+  });
+};
+
+const saveExecutedMigrationsFactory = (
+  client: Squidex<RestMigration>,
+) => async (migrations: string[]): Promise<void> => {
+  for (const migration of migrations) {
+    await client.create({
```

Comment:
This probably needs error handling & logging. If one create fails we don't want to lose the state of the migrations executed.

Reply:
This got me thinking. We probably don't want to run any further migrations if we can't store the state in the end. Probably we should instead migrate & save state for each iteration, instead of migrating all & save all. What to you think

Reply:
agreed

Reply:
as requested https://github.com/yldio/asap-hub/pull/784/commits/e3aedfdd8c716989ce66e9000602d18016ef7d30

sorry for force pushing but I rebased over master :D